# MA-GenTA

----------

## Abstract:

Current sequencing-based methods for profiling microbial communities rely on marker gene (e.g. 16S rRNA) or metagenome shotgun sequencing (mWGS) analysis. We present a quantitative, straightforward, cost-effective method for microbiome profiling that combines desirable features of both approaches termed MAâ€‘GenTA: Microbial Abundances from Genome Tagged Analysis. MA-GenTA employs highly multiplexed oligonucleotide probes designed from reference genomes in a pooled primer-extension reaction during library construction to derive relative abundance data. To test the utility of the MA-GenTA assay, probes were designed for 830 high quality metagenome-assembled genomes representing bacteria present in mouse stool specimens. Comparison of the MA-GenTA data with mWGS data demonstrated excellent correlation down to 0.01% relative abundance and a similar number of organisms detected per sample. Despite the incompleteness of the reference database, NMDS clustering based on the Bray-Curtis dissimilarity metric of sample groups was consistent between MA-GenTA, mWGS and 16S rRNA datasets.  MA-GenTA represents a potentially useful new method for microbiome community profiling based on reference genomes.

----------
## Overview:
There are 4 components of the MA-GenTA assay presented here. 

1. Probe design pipeline - this pipeline starts with reference genomes for probe design, filtering, and ends  with probe selection.

2. Probes used in MA-GenTA - these are the probes used in the MA-GenTA assay.

3. Data processing - starting with raw sequencing reads ending with count tables of MAGs and probes per sample.

4. Downstream analysis - starting with count tables and ending with statistical analyses and figure generation. 

